test/e2e/validation_test.go

```diff
@@ -334,3 +337,64 @@ func testExpectedServicesRunning(t *testing.T) {
 		})
 	}
 }
+
+// testCSRApproval tests if the BYOH CSR's have been approved by WMCO
+func testCSRApproval(t *testing.T) {
+	testCtx, err := NewTestContext()
+	require.NoError(t, err)
+	if gc.numberOfBYOHNodes == 0 {
+		t.Skip("BYOH CSR approval testing disabled")
+	}
+	for _, node := range gc.byohNodes {
+		csrs, err := testCtx.findNodeCSRs(node.GetName())
+		require.NotEqual(t, len(csrs), 0, "could not find BYOH node CSR's")
+		require.NoError(t, err, "could not find BYOH node CSR's")
+
+		for _, csr := range csrs {
+			isWMCOApproved := func() bool {
+				for _, c := range csr.Status.Conditions {
+					if c.Type != certificates.CertificateApproved || c.Reason != "WMCOApproved" {
+						return false
+					}
+				}
+				return true
+			}()
```

Comment:
Can CSRs be approved by multiple entities and still be considered valid for our use case? If so, then current logic does not hold.

Assume a CSR has 2 Conditions, `c1` and `c2`, in its status:
`c1.Type = certificates.CertificateApproved` and `c1.Reason != "SomeApproverThatIsNotWMCO"`
`c2.Type = certificates.CertificateApproved` and `c2.Reason != "WMCOApproved"`

Loop would return false (even if `c2` gets processed first) even though the CSR is WMCO approved.

```suggestion
			isWMCOApproved := func() bool {
				for _, c := range csr.Status.Conditions {
					if c.Type == certificates.CertificateApproved && c.Reason == "WMCOApproved" {
						return true
					}
				}
				return false
			}()
```

Reply:
Couldn't the CSR be approved by the `cluster-machine-approver` in a real world cluster? I am a bit worried that given we are scaling that deployment down to 0, we are never testing that scenario.

Reply:
I have accepted @saifshaikh48 suggestion on this. Additionally I don't think BYOH CSR's will be approved by cluster-machine-approver because the node to machine mapping will not be found which is used by cluster-machine-approver for validation.